[Wang Hongshu] iP

HongshuW opened the pull request:
# Duke ( Hongshu's version )
> “Your mind is for having ideas, not holding them.” – David Allen
#### *Duke is a program that helps you to better organise your tasks and time!*
## What can Duke do:
- [x] manage tasks
- [X] store a copy of the task list to your PC
- [X] read from the stored task lists when it starts running
## Features:
* add a task
  * todo {task description}
  * deadline {task description} /by yyyy-MM-dd HH:mm
  * event {task description} /at yyyy-MM-dd HH:mm to yyyy-MM-dd HH:mm
* delete a task: delete {task number}
* finish a task: done {task number}
* find a task: find {part of task description}
* list all tasks: list
* exit the program: bye
## All you need to do is :eyes::
1. download the jar file [here](https://github.com/HongshuW/ip/releases/download/v0.1/ip.jar)
2. double-click it
3. add your tasks
## If you Java programmer, you can use it to practice Java too.
Here's the ```main``` method:
```java
public static void main(String[] args) {
        String filePath = System.getProperty("user.dir");
        new Duke(Paths.get(filePath, "data", "duke.txt")).run();
}
```

Commit:
Add Gradle support

HongshuW committed:
Add Processor.java

HongshuW committed:
Duke class: change greeting message

HongshuW committed:
Processor class: add  method

HongshuW committed:
Duke class: add exiting message

HongshuW committed:
Duke class: run processor in a while loop

Duke class processes only one command and exits.

The program shouldn't exit after processing one command,
it should continue to run unless the 'bye' command is received.

Adding a while loop in `main` allows the program to keep running.
The return type of method `process` in class `Processor` is
changed from void to boolean, the method returns false only if
the command is 'bye'.

Changing the return type from void to boolean enables the while
loop to stop when 'bye' command is received.

HongshuW committed:
Rename Processor.java to Event.java

HongshuW committed:
Extract scanner from Event.java

Each event contains a scanner to read command.

This shouldn't be the case because an event should be triggered
after a command is received.

Scanner is removed from the Event class. A method called `readCommand`
is introduced in the Duke class, which will read the command and
decide the Duke program's next action.

Moving the reading command process into the Duke class enables
the program to decide which event to trigger based on the command.

HongshuW committed:
Rename Event.java to be Processor.java

HongshuW committed:
Add AddATaskProcessor.java

HongshuW committed:
Duke class: add  list

HongshuW committed:
Add GetListProcessor.java

HongshuW committed:
GetListProcessor class: implement process method

HongshuW committed:
Duke class: add getTasks method

HongshuW committed:
Add Task.java

HongshuW committed:
Duke class: change  into a list of s

HongshuW committed:
Task class: implement toString method

HongshuW committed:
Add TaskDoneProcessor.java

HongshuW committed:
Duke class: add done command

HongshuW committed:
Duke class: extract processCommand method

The steps of calling the process method of a processor and print
the result are repeated for each command.

These lines of code are repeated, which should be avoided.

In the Duke class, a method called processCommand is defined to
contain the repeated lines of code. Then the common lines in the
readCommand method are replaced by calling processCommand.

This makes it easier to debug and change the behavior later.

HongshuW committed:
Add Deadline.java

HongshuW committed:
Add ToDo.java

HongshuW committed:
Add Event.java

HongshuW committed:
Duke class: add three commands

HongshuW committed:
TextUiTesting: normal cases

HongshuW committed:
Add DukeException.java

HongshuW committed:
DukeException class: implement toString

HongshuW committed:
Duke class: handle exceptions

Invalid commands such as 'todo' and 'blah' are allowed, and these
will be treated as a todo task.

Allowing invalid commands may lead to problems such as recording a
typo as a todo task.

The following exceptions are handled using DukeException class:
- Invalid commands (commands other than bye, list, done, todo,
  deadline, event)
- Invalid task number in a done command
- Empty description in todo, deadline, event commands
- Empty time in deadline and event commands
- multiple time slots in deadline and event commands

By creating a new class, the error message can be changed into
the same format as other messages. Handling errors also provide
the users with a guideline on how to use the program.

HongshuW committed:
TextUiTesting: error handling test cases

HongshuW committed:
Add DeleteATaskProcessor.java

HongshuW committed:
Duke class: change tasks to ArrayList

HongshuW committed:
Duke class: add delete command

HongshuW committed:
TextUiTesting: test delete

HongshuW committed:
Add javadoc

HongshuW committed:
Duke class: add readFile and readData

HongshuW committed:
Task and subclasses: add toFileFormatString

HongshuW committed:
Duke class: write to file in addToList

HongshuW committed:
Duke class: write to file in removeFromList

HongshuW committed:
Duke class: add rewriteFile method

HongshuW committed:
Deadline class: parse time

HongshuW committed:
Event class: parse time

HongshuW committed:
Merge branch 'branch-Level-7'

HongshuW committed:
Merge branch 'branch-Level-8'

HongshuW committed:
Deadline and Event classes: fix time bug

HongshuW committed:
Add Ui.java

HongshuW committed:
Ui class: extract getCommand from Duke class

HongshuW committed:
Ui class: extract printing from Duke class

HongshuW committed:
Add Storage.java

HongshuW committed:
Add Parser.java

HongshuW committed:
Add ExitProcessor.java

HongshuW committed:
Add TaskList.java

HongshuW committed:
Duke class: change structure

HongshuW committed:
Extract Ui,Storage,Parser,TaskList

HongshuW committed:
Processor classes: rename to be Command

HongshuW committed:
Extract out exceptions

HongshuW committed:
Parser class: add parseCommandWithTime method

HongshuW committed:
Ui class: generalise methods into showError

HongshuW committed:
Duke class: change the logic of run method

HongshuW committed:
Split into packages

HongshuW committed:
TaskList class: follow tell-dont-ask principle

HongshuW committed:
JUnit test: Test Ui class

HongshuW committed:
JUnit test: test AddTaskCommand class

HongshuW committed:
Parser class: handle exceptions

HongshuW committed:
Add META-INF/MANIFEST.MF

HongshuW committed:
Add Javadoc

HongshuW committed:
Command class: add isRunning method

HongshuW committed:
Change variable name taskList to tasks

HongshuW committed:
Follow Java coding standard

HongshuW committed:
Add FindTaskCommand.java

HongshuW committed:
FindTaskCommand class: implement execute

HongshuW committed:
TaskList class: add filter method

HongshuW committed:
Parser class: enable parsing find command

HongshuW committed:
Merge branch 'branch-A-JavaDoc'

HongshuW committed:
Merge branch 'branch-A-CodingStandard'

HongshuW committed:
Merge branch 'branch-Level-9'

HongshuW committed:
Resolve conflicts

HongshuW committed:
Merge branch 'branch-Level-9'

HongshuW committed:
Merge branch 'branch-A-CodingStandard'

HongshuW committed:
Parser class: follow coding standard

HongshuW committed:
Parser class: fix bug

HongshuW committed:
Fix bug: task no

HongshuW committed:
DeleteTaskCommand class: fix deletion bug

HongshuW committed:
Merge remote-tracking branch 'origin/add-gradle-support'

HongshuW committed:
add checkstyle

HongshuW committed:
build.gradle: add dependencies

HongshuW committed:
Add MainWindow.java and MainWindow.fxml

HongshuW committed:
Add DialogBox.java and DialogBox.fxml

HongshuW committed:
Set up GUI

HongshuW committed:
GUI: add greeting message

HongshuW committed:
Duke class: implement getResponse method

HongshuW committed:
Duke class: stop extending Application

HongshuW committed:
MainWindow class: implement exit method

HongshuW committed:
Adjust gui

nzixuan commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,147 @@
+                this.readData(task);
+            }
+            fileScanner.close();
+        } catch (FileNotFoundException fileNotFoundException) { // If file doesn't exist, create it.
```
Should the comment be in a separate line from the code?
This is not explicitly stated in the Java coding standard but can help with consistency with the rest of the code.
Suggestion : 
```java
catch (FileNotFoundException fileNotFoundException) { 
    // If file doesn't exist, create it.
```

nzixuan replied:
Should the comment be in a separate line from the code?
This is not explicitly stated in the Java coding standard but can help with consistency with the rest of the code.

Suggestion :
```java
catch (FileNotFoundException fileNotFoundException) { 
    // If file doesn't exist, create it.
```

I have noticed this issue a several times in this file as well.

nzixuan commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,147 @@
+        } catch (FileNotFoundException fileNotFoundException) { // If file doesn't exist, create it.
+            try {
+                Storage.data.createNewFile();
+            } catch (IOException ioException) { // If directory doesn't exist, create it.
```
Refer to above comment on comment standards 
Suggestion: 
```java
            } catch (IOException ioException) { 
                // If directory doesn't exist, create it.

```

nzixuan commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,147 @@
+            } catch (IOException ioException) { // If directory doesn't exist, create it.
+                File directory = Paths.get(Storage.DIRECTORY_PATH, "data").toFile();
+                directory.mkdirs();
+                this.readFile(); // Run this method again to create a file.
```
Refer to above comment on comment standards
Suggestion:
```java
                directory.mkdirs();
                // Run this method again to create a file.
                this.readFile(); 
```

nzixuan reviewed (commented):
Good job overall with keeping with coding standards. It was difficult to find (if any) coding standard violations throughout the files. Well done!

HongshuW committed:
Perform checkstyle

jiayi1129 commented on src/main/java/duke/command/TaskDoneCommand.java:
```diff
@@ -0,0 +1,48 @@
+     * @param taskIndex Index of a task.
+     */
+    public TaskDoneCommand(int taskIndex) {
+        super("done");
```
Perhaps you can create a private static final variable for these magic literals? I noticed this issue in other places as well.

jiayi1129 commented on src/main/java/duke/exception/InvalidTaskNoException.java:
```diff
@@ -0,0 +1,13 @@
+package duke.exception;
+
+/**
+ * Represents an exception thrown when an invalid task number is given. A subclass of DukeException.
+ */
+public class InvalidTaskNoException extends DukeException {
+    /**
+     * Constructor of the class `InvalidTaskNoException`.
+     */
+    public InvalidTaskNoException() {
+        super("☹ OOPS!!! The task number is invalid.");
+    }
+}
```
I think it is really great that you have multiple types of `DukeException`! This makes understanding the exceptions thrown in this whole project more easily! 👍

jiayi1129 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,63 @@
+                + this.endingTime.format(DateTimeFormatter.ofPattern("MMM d yyyy, HH:mm")) + ")";
+    }
+
+    /**
+     * Converts the task to a string with the format of the file in hard disk.
+     *
+     * @return String representation of the task in the file's format.
+     */
+    @Override
+    public String toFileFormatString() {
+        return String.format("E / %s / %s / %s to %s\n", this.isDone ? "1" : "0", this.description,
+                this.startingTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")),
+                this.endingTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")));
+    }
```
Perhaps you can consider instantiating the logic for some of these string fields in variables first, then passing these variables into the `String.format` method?

```suggestion
    /**
     * Converts the task to a string with the format of the file in hard disk.
     *
     * @return String representation of the task in the file's format.
     */
    @Override
    public String toFileFormatString() {
        String doneIndicator = this.isDone ? "1" : "0";
        String startDateTime = this.startingTime.format(DateTimeFormatter.ofPattern(
                "yyyy-MM-dd HH:mm"));
        String endDateTime = this.endingTime.format(DateTimeFormatter.ofPattern(
                "yyyy-MM-dd HH:mm"));
        return String.format("E / %s / %s / %s to %s\n", doneIndicator, this.description, 
                startDateTime, endDateTime");
    }
```

jiayi1129 commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,123 @@
+        String regex = isEvent ? "/at" : "/by";
+        String taskType = isEvent ? "event" : "deadline";
+
+        // Parse the command
+        if (words.length >= 2) {
+            String[] information = words[1].split(regex);
+            if (information.length == 2) {
+                try {
+                    if (isEvent) {
+                        return new AddTaskCommand(new Event(information[0], information[1]));
+                    } else {
+                        return new AddTaskCommand(new Deadline(information[0], information[1]));
+                    }
+                } catch (DateTimeParseException | ArrayIndexOutOfBoundsException e) {
+                    throw new InvalidTimeException(timeFormat);
+                }
+            } else if (information.length < 2) {
+                throw new MissingCommandDetailException("time", taskType, String.format("%s %s", regex, timeFormat));
+            } else {
+                throw new MultipleTimeSlotsException(taskType);
+            }
+        } else {
+            throw new MissingCommandDetailException("description", taskType,
+                    String.format("%s %s", regex, timeFormat));
+        }
+    }
```
Perhaps you can try using switch case statement instead of if else + try catch conditionals to avoid deep nesting? Let me know what you think!

jiayi1129 commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,123 @@
+        // Split the command into two phrases
+        String[] words = command.split(" ", 2);
+
+        // Determine type of the command and return corresponding command instance
+        if (command.equals("bye")) {
+            return new ExitCommand();
+        } else if (command.equals("list")) {
+            return new GetListCommand();
+        } else if (words[0].equals("done")) {
+            return parseCommandWithTaskNo(words, true);
+        } else if (words[0].equals("delete")) {
+            return parseCommandWithTaskNo(words, false);
+        } else if (words[0].equals("todo")) {
+            if (words.length >= 2) {
+                return new AddTaskCommand(new ToDo(words[1]));
+            } else {
+                throw new MissingCommandDetailException("description", "todo", "");
+            }
+        } else if (words[0].equals("deadline")) {
+            return Parser.parseCommandWithTime(words, false);
+        } else if (words[0].equals("event")) {
+            return Parser.parseCommandWithTime(words, true);
+        } else if (words[0].equals("find")) {
+            if (words.length >= 2) {
+                return new FindTaskCommand(words[1].trim());
+            } else {
+                throw new MissingCommandDetailException("keyword", "find", "");
+            }
+        } else {
+            throw new InvalidCommandException();
+        }
```
Perhaps you can consider using switch case statements?

jiayi1129 reviewed (commented):
Overall, I like your program! Other than the nested if else/try catch conditionals, the other parts were well structured and well commented! 👍

Other than a few minor nitpicks, LGTM! 😄

HongshuW committed:
build.gradle: change main class to be Launcher

HongshuW committed:
Assert task to be not null

Task objects and methods are used in various classes.

If a task object is null, its methods cannot be called and there
will be a nullPointerException.

Therefore task objects are asserted so that they cannot be null.

Assertions will not affect the program but they will notify the
programmer that the task shouldn't be null.

HongshuW committed:
Change some variables to final

HongshuW committed:
Parser class: reduce the level of nesting

HongshuW committed:
Parser class: reduce nesting further more

HongshuW committed:
Parser class: split long methods into shorter ones

HongshuW committed:
MainWindow class: remove unused variable

HongshuW committed:
TaskList class: filter list by using stream

ArrayList tasks is filtered by a for loop, then the filtered task
list will be converted to strings and concatenated to the respond
message in FindTaskCommand and GetListCommand classes.

This violates the tell-don't-ask principle and this process should
be done in the TaskList class.

In the TaskList class, getFilteredListAsString method converts the
task list to a string. The array list is converted to a stream,
filtered, then converted to a string.

In such a way, the command classes can ask for the final string
instead of ask for the list and then convert it to string.

HongshuW committed:
TaskList class: add javadoc

HongshuW committed:
Merge pull request #1 from HongshuW/branch-A-Assertions

Assert task to be not null

HongshuW committed:
Merge branch 'master' into branch-A-CodeQuality

HongshuW committed:
Merge branch 'master' into branch-A-Streams

HongshuW committed:
Merge pull request #2 from HongshuW/branch-A-CodeQuality

Branch-a-code-quality

HongshuW committed:
Merge branch 'master' into branch-A-Streams

HongshuW committed:
Merge pull request #3 from HongshuW/branch-A-Streams

Branch-a-streams

HongshuW committed:
Create gradle.yml

HongshuW committed:
Remove unused imports

HongshuW committed:
Update gradle.yml

HongshuW committed:
Add TaskUndoneCommand.java

HongshuW committed:
Add UnableToUndoException.java

HongshuW committed:
Add UndoCommand.java

HongshuW committed:
Parser class: parse undo command

HongshuW committed:
Merge branch 'branch-C-Undo'

HongshuW committed:
Handle InvalidTaskNoException

HongshuW committed:
Add HelpCommand.java

HongshuW committed:
Duke class: strip input before parsing

HongshuW committed:
Rename Duke to be WorkDone

HongshuW committed:
Rename package duke to be package workdone

HongshuW committed:
Rename DukeException to WorkDoneException

HongshuW committed:
Checkstyle and improve code

HongshuW committed:
DialogBox class: change dialog colors

HongshuW committed:
Command class: change response message format

HongshuW committed:
WorkDoneException class: change error message format

HongshuW committed:
Main class: add a title to the stage

HongshuW committed:
Add ClearCommand.java

HongshuW committed:
Add Ui.png

HongshuW committed:
Set theme jekyll-theme-merlot

HongshuW committed:
Set theme jekyll-theme-midnight

HongshuW committed:
Set theme jekyll-theme-midnight

HongshuW committed:
README.md: update user guide

HongshuW committed:
Merge branch 'master' of https://github.com/HongshuW/ip

HongshuW committed:
MainWindow class: remove alert box
